### PR TITLE
TINY-9654: Update Robin's word identification logic to identify words containing non-Latin characters

### PR DESCRIPTION
**Pull request by mkzhx**
Related Ticket: TINY-9654

Description of Changes:

In Polaris:
* Adapt `Words.getWords` logic to create new API `Words.getWordsWithIndices` to also return start and end indices
* Removed obsolete TINY-1166 fix that stripped all `\uFEFF` characters in `Words.getWords`
* Update `WordBoundary.isWordBoundary`  logic for extend and format characters to address the following examples returning no words in `Words.getWords`:
    * "a`\u0300`b" (extend character in word with a single character on either side)
    * "a`\ufeff`b" (format character in word with a single character on either side)
    * "a`\ufeff` " (format character that also counts as whitespace preceding a word boundary)

In Robin:
* Make `Identify.words` use Polaris's `Words.getWordsWithIndices`
* `Identify.words` is used by Robin's `Words.identify`, `TextZones.single`, and `TextZones.range` APIs, so their behaviours are updated

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [x] Milestone set
* [x] ~Docs ticket created (if applicable)~

GitHub issues (if applicable):

### CODE REVIEW

File: modules/polaris/src/main/ts/ephox/polaris/words/WordBoundary.ts [hunk 1]
```diff
@@ -56,9 +56,16 @@ const isWordBoundary = (map: CharacterMap, index: number): boolean => {
...
+  if ((type === ci.EXTEND || type === ci.FORMAT) &&
+    (nextType === ci.ALETTER || nextType === ci.NUMERIC || nextType === ci.KATAKANA ||
+      nextType === ci.EXTEND || nextType === ci.FORMAT)
+    ||
+    (nextType === ci.EXTEND ||
+      // TINY-9654: Only ignore format characters if they do not precede a word boundary. Since some format characters overlap with whitespace characters (ex: \ufeff) and
+      // our word extraction logic excludes whitespace characters, if a whitespace-overlapping format character that precedes a word boundary is not split on, whichever word
+      // it is a part of will not be added to the list of extracted words, causing inaccuracies.
+      nextType === ci.FORMAT && (nextNextType === ci.ALETTER || nextNextType === ci.NUMERIC || nextNextType === ci.KATAKANA || nextNextType === ci.EXTEND || nextNextType === ci.FORMAT))
+      && (type === ci.ALETTER || type === ci.NUMERIC || type === ci.KATAKANA || type === ci.EXTEND || type === ci.FORMAT)) {
```
Comment (TheSpyder) [inline modules/polaris/src/main/ts/ephox/polaris/words/WordBoundary.ts:68]: This is a bit complex to review in a PR so I'm going to trust that we have automated tests for this file, and none of them broke (other than `'twas` which is an improvement) so this doesn't regress anything.

File: modules/polaris/src/main/ts/ephox/polaris/words/Words.ts [hunk 2]
```diff
@@ -26,9 +26,19 @@ const findUrlEnd = (characters: string[], startIndex: number): number => {
   return peakedWord.substr(0, 3) === '://' ? endIndex : startIndex;
 };
 
-const findWords = <T>(chars: T[], sChars: string[], characterMap: CharacterMap, options: WordOptions): T[][] => {
-  const words: T[][] = [];
-  let word: T[] = [];
+export type Word<T> = T[];
+interface WordIndex {
+  start: number;
+  end: number;
```
Comment (TheSpyder) [inline modules/polaris/src/main/ts/ephox/polaris/words/Words.ts:32]: I always forget this is necessary too.

```suggestion
  readonly start: number;
  readonly end: number;
```

File: modules/polaris/src/main/ts/ephox/polaris/words/Words.ts [hunk 3]
```diff
@@ -26,9 +26,19 @@ const findUrlEnd = (characters: string[], startIndex: number): number => {
...
+}
+export interface WordsWithIndices<T> {
+  words: Word<T>[];
+  indices: WordIndex[];
```
Comment (TheSpyder) [inline modules/polaris/src/main/ts/ephox/polaris/words/Words.ts:36]:
```suggestion
  readonly words: Word<T>[];
  readonly indices: WordIndex[];
```

**Review by TheSpyder — APPROVED** [review]
Looking good! Well done!

File: modules/polaris/src/main/ts/ephox/polaris/words/WordBoundary.ts (hunk 1 above)
Comment (mkzhx) [reply modules/polaris/src/main/ts/ephox/polaris/words/WordBoundary.ts:68]: The `'twas` change is a collateral improvement after switching to Polaris's `getWords` API. 

The above change with the word boundary logic for extend and format characters addresses a bug where while "aa`\ufeff`bb" is considered a word by `getWords`, while "a`\ufeff`b" is not considered a word. Format characters like `\ufeff` were supposed to be ignored, so "a`\ufeff`b" should be 1 word.

The bug was uncovered after I removed the obsolete TINY-1166 fix that stripped `\ufeff` characters in `getWords` (`Identify.words` requires `\ufeff` characters to be retained as per https://github.com/tinymce/tinymce/blob/d6e308d85ec9b1bf9285b11f5bb6d0cf19ed0e95/modules/robin/src/test/ts/browser/zone/EntityTest.ts#L24)

File: modules/polaris/src/main/ts/ephox/polaris/words/Words.ts [hunk 4]
```diff
@@ -26,9 +26,19 @@ const findUrlEnd = (characters: string[], startIndex: number): number => {
...
+export type Word<T> = T[];
+interface WordIndex {
+  readonly start: number;
+  readonly end: number;
+}
+export interface WordsWithIndices<T> {
+  readonly words: Word<T>[];
+  readonly indices: WordIndex[];
+}
+const findWordsWithIndices = <T>(chars: Word<T>, sChars: string[], characterMap: CharacterMap, options: WordOptions): WordsWithIndices<T> => {
```
Comment (spocke) [inline modules/polaris/src/main/ts/ephox/polaris/words/Words.ts:38]: I'm told to not be picky about formatting but maybe space this up a bit.

```suggestion
export type Word<T> = T[];

interface WordIndex {
  readonly start: number;
  readonly end: number;
}

export interface WordsWithIndices<T> {
  readonly words: Word<T>[];
  readonly indices: WordIndex[];
}

const findWordsWithIndices = <T>(chars: Word<T>, sChars: string[], characterMap: CharacterMap, options: WordOptions): WordsWithIndices<T> => {
```

**Review by spocke — APPROVED** [review]

**Pull request merged**